src/single-spa-alpinejs.test.js

```diff
@@ -1,17 +1,269 @@
+        </template>
+        <!-- end: user card -->
+        </div>
+    </div>
+  `;
+
+  let props;
+  let propsTwo;
+
+  beforeEach(() => {
+    const div = document.createElement("div");
+    div.id = "test-div";
+    document.body.appendChild(div);
+
+    props = { name: "test" };
+    propsTwo = { name: "testTwo" };
+  });
+
+  afterEach(() => {
+    document.getElementById("test-div").remove();
+    window.xInitFn = null;
+  });
+
   it(`Can create lifecycle functions`, () => {
     const lifecycles = singleSpaAlpinejs({
-      template: `
-        <div x-show="open" class="mui--text-display4">
-            Hey, I'm open
-        </div>
-      `,
+      template: appOneTemplate.trim(),
     });
 
+    const domEl = domElementGetter();
+    expect(domEl.innerHTML.trim()).toBe("");
     expect(typeof lifecycles.bootstrap).toBe("function");
     expect(typeof lifecycles.mount).toBe("function");
     expect(typeof lifecycles.unmount).toBe("function");
   });
+
+  it(`Can create lifecycle functions with x-data as object`, () => {
+    const lifecycles = singleSpaAlpinejs({
+      template: appTwoTemplate.trim(),
+      xData: { open: false },
+    });
+
+    const domEl = domElementGetter();
+    expect(domEl.innerHTML.trim()).toBe("");
+    expect(typeof lifecycles.bootstrap).toBe("function");
+    expect(typeof lifecycles.mount).toBe("function");
+    expect(typeof lifecycles.unmount).toBe("function");
+  });
+
+  it(`Can create lifecycle functions with x-data as function`, () => {
+    const lifecycles = singleSpaAlpinejs({
+      template: appTwoTemplate.trim(),
+      xData: () => ({ open: false }),
+    });
+
+    const domEl = domElementGetter();
+    expect(domEl.innerHTML.trim()).toBe("");
+    expect(typeof lifecycles.bootstrap).toBe("function");
+    expect(typeof lifecycles.mount).toBe("function");
+    expect(typeof lifecycles.unmount).toBe("function");
+  });
+
+  it(`throws if we create lifecycle functions with x-data as string`, () => {
+    expect(() =>
+      singleSpaAlpinejs({
+        template: appTwoTemplate.trim(),
+        xData: "({ open: false })",
+      })
+    ).toThrow();
+  });
+
+  it(`throws if you don't provide a template`, () => {
+    expect(() => {
+      singleSpaAlpinejs({});
+    }).toThrow();
+  });
+
+  it(`throws if you provide a non-string template`, () => {
+    expect(() => {
+      singleSpaAlpinejs({
+        template: 123,
+      });
+    }).toThrow();
+  });
+
+  it(`throws if you provide a domElementGetter that is not a function`, () => {
+    expect(() => {
+      singleSpaAlpinejs({
+        template: {},
+      });
+    }).toThrow();
+  });
+
+  it(`renders function template with x-data , template and props`, () => {
+    const lifecycles = singleSpaAlpinejs({
+      template: () => Promise.resolve(appTwoTemplate.trim()),
+      xData: { open: false },
+      domElementGetter,
+    });
+    const domEl = domElementGetter();
+    return lifecycles
+      .bootstrap(props)
```

Comment:
This test is great. Here's one tip with jest tests that I find handy - feel free to ignore/not change the code as there is no advantage to this approach over what you have. But you can make jest tests async functions and then use await for a bit of syntax sugar:

```js
it(`renders function template with x-data , template and props`, async () => {
  await lifecyles.bootstrap(props);
  await lifecycles.mount(props);
})
```

If you run into regeneratorRuntime issues, try setting the BABEL_ENV inside of the package.json:

https://github.com/single-spa/single-spa-alpinejs/blob/c7f720d151d4bedc3e589c3b54814f52d686fa54/package.json#L9

Example:

```json
{
  "scripts": {
    "test": "cross-env BABEL_ENV=test jest"
  }
}
```

See https://github.com/kentcdodds/cross-env and https://babeljs.io/docs/en/6.26.3/babelrc#env-option. Also https://github.com/single-spa/single-spa-alpinejs/blob/c7f720d151d4bedc3e589c3b54814f52d686fa54/.babelrc#L4

Reply:
Thats great insight Joel .. After adding some more tests I found that I was getting a Promise rejection error , so converted those tests to use async and they got resolved

Reply:
added this